### DISCUSSION TITLE
Granularidade de componentes, até que ponto?

### DISCUSSION THREAD

**Discussion by thulioph** [frontendbr/forum#499] [Discussão]
Gostaria de trazer pra cá, uma discussão sobre até que ponto vale a pena a granularização de um determinado componente. 

> Exemplo:

- Acabei de construir um componente que possui uma determinada marcação, estilo e comportamento.
- Uma outra pessoa que pega o componente, diz que aquele componente poderia ser separado em mais 2 componentes menores.

Durante o desenvolvimento, normalmente eu avalio:

**Principío da responsabilidade única:**

Aquele meu componente, irá conter marcação e estilo para resolver o problema que ele foi destinado a solucionar, e qualquer lógica fora da que o componente propõe, deve ser utilizada em outro arquivo.

**Reaproveitamento:**

Esse componente será reaproveitado em outro lugar da aplicação? Ele pode ser quebrado em partes menores que façam sentido e não deixam o código inflado? 

Quais seriam outros fatores para se avaliar como tomada de decisão? Ou como ter argumentos pra se discutir com sanidade sobre essa granularidade? Apenas **reaproveitamento** e **pru** resolvem?

**Comment by nandomoreirame** [comment]
Good question!

**Comment by nandomoreirame** [comment]
Existe esse mesmo questionamento aqui no trabalho quanto APIs e Microservices.

**Comment by vinicius73** [comment]
Levo sempre em consideração a quantidade de lógica que aquele componente carrega.
Nem todo componente é realmente reaproveitável em outro contexto, porém não fragmentar prejudica a manutenção.
Como uso programação funcional, levo isso para meus componentes. 
Em alguns casos fragmento pedaços de lógica do componente, pedaços que nem refletem nele visualmente, tiro proveito dos mixins do Vue.js para criar "pedaços de lógica" dinamicamente


```js
export default {
  mixins: [ mapSetupFields({
    location: 'setup.location',
    country: 'setup.country',
    city: 'setup.city',
    region: 'setup.region'
  }) ]
}
```

Se tratando de programação, quase toda abstração é bem vinda, inclusive na hora de criar componentes visuais. Mesmo assim sempre deve haver um "limite"... há algumas fragmentações que mais atrapalham do que ajudam.

**Comment by matheus-vieira** [comment]
Na minha ínfima opnião creio que o limite  é o contexto atual, ou seja, não acho nada produtivo ficar pensando em hipóteses "ah se eu fizer assim então eu vou, quem sabe, utilizar" isso seria um argumento muito vago.

Primeiramente resolva seu problema! Claro que sempre de forma mais simples e clara possível.
"Segundamente" se tem alguma refatoração em mente não precisa implementar no momento, deixe um comentário explicado sua "melhor solução" para quando o código for revisitado verificar se ainda há sentido na sua lógica.

Espere um problema aparecer para resolver, não tente abraçar o mundo.

**Comment by mtmr0x** [comment]
@thulioph eu nao sei o quanto de tamanho cada componente tem ou qual sua responsabilidade e sua markup, mas granularidade contribui em:

 * testes unitarios;
 * escalabilidade dos componentes;
 * reaproveitamento;

Tu disse sobre *PRU*, o que eh bem legal, e falaram sobre resolver o seu problema na hora, mas falar em escalar e resolver o problema na hora sao contraditorios, imagino eu, fortemente, sobre isso.

Se seus testes ficariam mais simples com eles mais granulados, eu, Matheus, os separariam, sugiro o mesmo, realmente facilita, mas eu tbm nao estou vendo seu caso especificamente, so dando meus 2 cents aqui.

bjos de luz
